rfcs/convergence/internationalization.md

```diff
@@ -0,0 +1,217 @@
+
+- Easily propagate app-level locale strings down to Fluent components without needing to manually map them to multiple separate component props
+- Easily include all necessary component strings when they create a new locale definition/JSON file (i.e. not need to manually remember that TagPicker needs `selectionAriaLabel` defined).
+- Use 3rd party i18n tools with Fluent components without too much extra work
+
+To do this, I think it makes sense to add locale data to Fluent's `ProviderContext`, since it already includes `dir`.
+
+One straightforward option for how to do this would be to provide a locale object with static key/value pairs for each string. We could provide a minimal set of global strings in the top-level provider, similar to how Material and Ant differentiate global vs. component strings. Based on current Fluent v8 and N\* strings, it would look like this:
+
+react-shared-contexts/src/ProviderContext/defaultStrings.ts
+
+```js
+export const defaultLocale {
+  locale: 'en-US',
+  strings: {
+    global: {
+      more: 'more items',
+      close: 'close',
+      noResults: 'no results found',
+      loading: 'loading',
+      selectAll: 'select all',
+      expandCollapse: 'expand {0}'
+    }
+  }
+}
+```
+
+react-shared-contexts/src/ProviderContext/ProviderContext.ts
+
+```js
+import { defaultLocale } from './defaultStrings.ts';
+
+export const ProviderContext =
+  React.createContext <
+  ProviderContextValue >
+  {
+    targetDocument: typeof document === 'object' ? document : undefined,
+    dir: 'ltr',
+    locale: defaultLocale,
+  };
+```
+
+Then, each component would take a `strings` prop where strings could be directly passed in, and also provide component-specific defaults when needed. The `strings` prop would also be used for the author to pass in strings that don't have a default, e.g. error messaging and instance-specific labels.
+
+react-dropdown/src/localeStrings.ts (specific strings are just for this example)
+
+```js
+export const dropdownDefaultStrings: Partial<DropdownStrings> {
+  loadMore: 'Load more options',
+  invalidEntryError: 'Your search did not match any options'
+}
+```
+
+react-dropdown/src/components/useDropdown.tsx (very simplified)
+
+```js
+import { dropdownDefaultStrings } from '../localeStrings';
+export const useDropdown = props => {
+  const { locale } = useFluent();
+  const dropdownStrings = {
+    ...locale.strings.global, // only if this component actually uses global strings
+    ...locale.strings.Dropdown,
+    ...props.strings,
+  };
+
+  state.components.arrow['aria-label'] = dropdownStrings.expandCollapse;
+};
+```
+
+## Concrete actions for vNext beta
+
+The only action needed for beta release should be to ensure all strings provided through component props (excluding children/child content) are defined in a single `strings` property.
```

Comment:
`strings` is not the best name for localisation purposes: there are cases when it should be actually functions. Especially when it goes to plurals for different languages:

```ts
<Component
  strings={{
    getApples: count => {
      if (count === 0) return "No apples";
      return `${count} apple(s)`;
    }
  }}
/>
```

Is it good idea to have functions under strings? Or there will be two props?

```ts
<Component
  strings={{ busy: 'Busy' }}
  functions={{
    getApples: count => {
      if (count === 0) return "No apples";
      return `${count} apple(s)`;
    }
  }}
/>
```

Reply:
Having a prop for localization can also play a bad game with consistency in a big application where components are not localized consistently and each application area has own strings that don't match or are missing the proper translation.

---

The value `strings` prop is also questionable when it comes to different languages. Do we really expect to have this?

```tsx
function App() {
  const locale = useLocale();
  return (
    <List>
      <ListItem strings={SOME_SHARED_STRINGS[locale]} />
      <ListItem strings={SOME_SHARED_STRINGS[locale]} />
      <ListItem strings={SOME_SHARED_STRINGS[locale]} />
      <ListItem strings={SOME_SHARED_STRINGS[locale]} />
      {/* 50 times more */}
    </List>
  );
}
```

Reply:
As we don't have an agreement and complete design, but it have been already leaked to existing components (https://github.com/microsoft/fluentui/pull/22775) to unblock localization I propose at least prefix it with `unstable_` to have some room for changes.